h2o-r/h2o-package/R/frame.R

```diff
@@ -3228,34 +3228,53 @@ as.h2o.Matrix <- function(x, destination_frame="", ...) {
   .key.validate(destination_frame)
   if ( destination_frame=="" ) # .h2o.readSVMLight wont handle ""
     destination_frame <- .key.make("Matrix") # only used if `x` variable name not valid key
-  
-  tmpf <- tempfile(fileext = ".svm")
-  .h2o.write.matrix.svmlight(x, file = tmpf)
-  h2f <- .h2o.readSVMLight(tmpf, destination_frame = destination_frame)
-  file.remove(tmpf)
-  h2f
+
+  drs = as.simple_triplet_matrix(x)# need to convert sparse matrix x to a simple triplet matrix format
+  thefile <- tempfile()
+  .h2o.write_stm_svm(drs, file = thefile)
+  h2f <- h2o.uploadFile(thefile, parse_type = "SVMLight")
+  unlink(thefile)
+  h2f[- c(1)]# remove the first column
+
 }
 
-.h2o.write.matrix.svmlight <- function(matrix, file) {
-  on.exit(sink())
-  sink(file)
-  sapply(1:nrow(matrix), function(i) {
-    r <- matrix[i, ]
-    val.indices <- which(r != 0)
-    val.indices <- val.indices[val.indices > 1]
-    target <- r[1]
-    features <- paste(sprintf("%d", val.indices - 1), r[val.indices], collapse = " ", sep = ":")
-    line <- sprintf("%s %s\n", target, features)
-    cat(line)
-  })
+# Convert a simple triplet matrix to svm format
+# @author Peter Ellis
+# @return a character vector of length n = nrow(stm)
+.h2o.calc_stm_svm <- function(stm, y){
+  # returns a character vector of length y ready for writing in svm format
+  if(!"simple_triplet_matrix" %in% class(stm)){
+    stop("stm must be a simple triple matrix")
+  }
+  if(!is.vector(y) | nrow(stm) != length(y)){
+    stop("y should be a vector of length equal to number of rows of stm")
+  }
+  n <- length(y)
+
+
+  # data.table solution thanks to roland
+  rowLeft <- setdiff(c(1:n), unique(stm$i))  # added two liner to return rows of zeros instead of repeating other rows
+  nrowLeft <- length(rowLeft)
+
+  stm2 <- data.table(i = c(stm$i,rowLeft), j = c(stm$j,rep(1,nrowLeft)), v = c(stm$v,rep(0,nrowLeft)))
```

Comment:
@michalkurka Yes, good catch.  @wendycwong You can use the example here to wrap this in an if statement: https://github.com/h2oai/h2o-3/blob/98f6123b99875b1e00bc1eb7dd24afde559a0af1/h2o-r/h2o-package/R/frame.R#L3205